onnxruntime/core/graph/graph.cc

```diff
@@ -993,7 +975,11 @@ void Node::AddAttribute(std::string attr_name, GraphProto value) {
 bool Node::ClearAttribute(const std::string& attr_name) {
   graph_->SetGraphResolveNeeded();
   graph_->SetGraphProtoSyncNeeded();
-  return attributes_.erase(attr_name) > 0;
+  size_t erased = attributes_.erase(attr_name);
+  if (erased && attr_name == kBackwardNodeAttributeName) {
+    isForwardNode_ = true;
+  }
+  return erased > 0;
```

Comment:
I'm not familiar with training as well. IMHO it doesn't matter if this attribute gets removed or not in normal workflows. What matters is that it can be removed and not checking for removal would change the behaviour and potentially even introduce bugs.

Reply:
In this case it doesn't seem like a standard attribute, so I would like to understand how it's used before adding more stuff to the production code to handle it theoretically changing. There's a binary size and maintenance cost, and if it's a purely internal value with specific usage it may be better to validate the usage remains as expected via unit tests.

Based on [this](https://github.com/search?q=repo%3Amicrosoft%2Fonnxruntime%20kBackwardNodeAttributeName&type=code) github code search it has a very internal name, only seems to be set in a training optimizer, and only seems to be read in the graph_viewer code.

Unless there's some external usage of this magic value outside of ORT, it seems like it would be simpler for a Node to have a bool member that is directly set by the training optimizer instead of the indirect costly usage of a specially named attribute. 

@askhade do you know if this special value is used outside of ORT and must be set in the Node attributes?

Reply:
I checked the code and IMO it is OK to make this a book member of the Node class instead of making it a named attribute. There is widespread usage of this in the code so the change should not be very cumbersome. @mtavenrath let me know if you have any questions regarding this. Tagging @pengwa to validate this. 

@mtavenrath what timeline are you targeting for this change? We may need a couple of days to get this reviewed from Peng.

Reply:
One note - when finding usage to update you need to search for both the kBackwardNodeAttributeName constant as well as the string "__backwardpass". Ideally we can make all places use the constant.

https://github.com/search?q=repo%3Amicrosoft%2Fonnxruntime%20kBackwardNodeAttributeName&type=code
https://github.com/search?q=repo%3Amicrosoft%2Fonnxruntime%20__backwardpass&type=code

Reply:
> I checked the code and IMO it is OK to make this a book member of the Node class instead of making it a named attribute. There is widespread usage of this in the code so the change should not be very cumbersome. @mtavenrath let me know if you have any questions regarding this. Tagging @pengwa to validate this.
> 
> @mtavenrath what timeline are you targeting for this change? We may need a couple of days to get this reviewed from Peng.

FYI @askhade @skottmckay Yes, it is a purely internal used attributes, and 
1. it is firstly introduced as a backward-data-range specific perf improvement in https://github.com/microsoft/onnxruntime/blame/dfeda9019cfed2d6df5bcacc54269c7de481bdee/onnxruntime/core/providers/rocm/rocm_kernel.h#L29.  

3. A second usage pattern is: in priority based topo ordering, we consider the backward tagged node has lower priority than forward node, in training code path https://github.com/microsoft/onnxruntime/blob/dfeda9019cfed2d6df5bcacc54269c7de481bdee/orttraining/orttraining/core/optimizer/memory_optimizer/memory_insight.cc#L172. 

While both usage are having similar logic when set the backward pass to be true, e.g:

"
  for (auto node_index : node_topology_list) {
    auto& node = *graph.GetNode(node_index);

    if (node.OpType() == "YieldOp") {
      is_backward_pass = true;
    }
"
in https://github.com/microsoft/onnxruntime/commit/d5d69246885befb4c52ee9deef8946121834a00b#diff-8d8d103ec215ba8edb8ab23e876080adfd60f6f377084ffeca041c8b4f189a2cR13

and 

"
// Find the YieldOp node.
  Node* yield_op_node = nullptr;
  for (auto& node : graph.Nodes()) {
    if (node.OpType() == "YieldOp") {
      yield_op_node = &node;
      break;
    }
  }
" in https://github.com/microsoft/onnxruntime/blob/dfeda9019cfed2d6df5bcacc54269c7de481bdee/orttraining/orttraining/core/optimizer/memory_optimizer/memory_insight.cc#L172


Plus "YieldOp" is used in ORTModule (e.g. --enable_training) build only. So I think it is possible we restrict the getforward/setforward in ENABLE_TRAINING macro. While this may need some change in [onnxruntime/core/session/provider_bridge_ort.cc](https://github.com/microsoft/onnxruntime/blob/dfeda9019cfed2d6df5bcacc54269c7de481bdee/onnxruntime/core/session/provider_bridge_ort.cc#L885) to wrap the new bool property ( Not sure whether your tried build/running your local code with ROCM, while I feel the change should be needed to make the ROCM ep code work.).

Reply:
One thing I noticed for a while. Both training and inference build can use priority based order, while it is indeed needed for some training features, while I don't know how much value it has for model inferencing. 

If most inference users don't have such a need, maybe we can load the `nodes_in_topological_order_with_priority_`  in lazy mode, e.g. we only initialize it when first time user needed it via ```GetNodesInTopologicalOrder(ExecutionOrder::PRIORITY_BASED)``` . 

@skottmckay  @askhade